sessionctx/variable/sysvar.go

```diff
@@ -1838,7 +1838,23 @@ var defaultSysVars = []*SysVar{
 		s.EnablePseudoForOutdatedStats = TiDBOptOn(val)
 		return nil
 	}},
-
+	{Scope: ScopeGlobal | ScopeSession, Name: TiDBRestrictCharacterMode, Value: collate.DefaultCharMode, Validation: func(vars *SessionVars, normalizedValue string, originalValue string, scope ScopeFlag) (string, error) {
+		// Ensure the SQL mode parses
+		normalizedValue = collate.FormatCharMode(normalizedValue)
+		if _, invalidMode := collate.GetCharMode(normalizedValue); invalidMode != "" {
+			return originalValue, ErrWrongValueForVar.GenWithStackByArgs(TiDBRestrictCharacterMode, invalidMode)
+		}
+		return normalizedValue, nil
+	}, SetSession: func(s *SessionVars, val string) error {
+		val = collate.FormatCharMode(val)
+		charMode, invalidMode := collate.GetCharMode(val)
+		if invalidMode != "" {
+			return ErrWrongValueForVar.GenWithStackByArgs(TiDBRestrictCharacterMode, invalidMode)
+		}
+		s.SkipUTF8Check = charMode.SkipUTF8Check()
+		s.EnableInvalidCharacter = charMode.EnableInvalidCharacter()
+		return nil
+	}},
```

Comment:
This is adding a new sysvar of type string. We have `TypeEnum`, but we don't have `TypeSet`. I think this could be useful, as we can then use the code to generate `Possible Values` in the docs like this: https://docs.pingcap.com/tidb/stable/system-variables#default_authentication_plugin

May I suggest adding it?

You would still need to split the string in the GetSession func, but you could rely on the TypeSet to provide all the validation.

Reply:
it looks great, I add the TypeSet.